pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/Designer.java

```diff
@@ -10,10 +10,13 @@
 import java.util.Objects;
 import java.util.stream.Collectors;
 
+import org.apache.commons.lang3.ArrayUtils;
+
 import net.sourceforge.pmd.PMDVersion;
 import net.sourceforge.pmd.util.fxdesigner.app.DesignerRoot;
 import net.sourceforge.pmd.util.fxdesigner.util.DesignerUtil;
 
+import com.sun.javafx.fxml.builder.ProxyBuilder;
```

Comment:
complete ignorance here… but is it ok to import this from `com.sun`? can it break under non-oracle JREs?

Reply:
The internal API of OpenJFX is hosted in the subpackages of `com.sun.javafx`. So, this package is not Oracle-specific,  it's javafx-specific. It would break only if there's no JavaFX distribution on the classpath, in which case the whole app would naturally be unlaunchable. It would also break if OpenJFX decides to change the package name but that seems unlikely.

But since you mention it, I haven't tested whether that works on JREs other that 9. Maybe the class has been moved, since it's internal API. That has happened a lot with control implementations. I'll check

Reply:
Maybe also there's a published builder class somewhere, I'll dig further

Reply:
So I couldn't find a way to use only published API, since it's very limited. The ProxyBuilder class is still internal on the [latest trunk](https://hg.openjdk.java.net/openjfx/jfx-dev/rt/file/0ee0cd385816/modules/javafx.fxml/src/main/java/com/sun/javafx/fxml/builder/ProxyBuilder.java) of OpenJFX, but hasn't moved or been renamed, so we should be good.

Reply:
It seems to be still working with java8 and java9 - I would have expected some problems, since there javafx is run as a module and "com.sun.javafx.fxml.builder" is not exported (https://hg.openjdk.java.net/openjfx/jfx-dev/rt/file/0ee0cd385816/modules/javafx.fxml/src/main/java/module-info.java).
With openjdk and java11, javafx is just on the classpath and we can access all of it. If ProxyBuilder is removed, we'll notice it as a compile time error. Of course, we don't have full control about the runtime, since we don't ship javafx included - maybe we should do this? (later). But as you said, it's unlikely that this class would be removed, since it seems to be central for loading FXML and instantiating the referenced classes.

But looking where this class is used in javafx.fxml, I found this one: [JavaFXBuilderFactory](https://hg.openjdk.java.net/openjfx/jfx-dev/rt/file/0ee0cd385816/modules/javafx.fxml/src/main/java/javafx/fxml/JavaFXBuilderFactory.java). 

This seems to be working, but we still need to downcast (or use reflection, which would move the problem to runtime error....):

``` java
                JavaFXBuilderFactory builderFactory = new JavaFXBuilderFactory();
                Builder<?> builder = builderFactory.getBuilder(type);
                ((ProxyBuilder<?>) builder).put("designerRoot", owner);
                return builder;
```

Reply:
Right, we could also use JavaFXBuilderFactory and downcast to `Map<String,Object>`, so as not to use ProxyBuilder explicitly. But then as you said, it's probably better to get a compile error here than an NPE at runtime in some random control which hasn't been injected.

In fact I'm quite surprised that that works on Java 9 even with no `--add-opens ..=ALL_UNNAMED` since the package is not exported...

Maybe we should ship OpenJFX yes, but probably when figure out how we'll distribute the designer separately

Reply:
I tested it again and it works with all version (openjdk11, openjdk10, oracle jdk 11, oracle jdk 10, oracle jdk 9, oracle jdk 8). When running with oracle jdk 8, I get the following warning:

```
Feb 23, 2019 7:29:08 PM javafx.fxml.FXMLLoader$ValueElement processValue
WARNUNG: Loading FXML document with JavaFX API of version 9 by JavaFX runtime of version 8.0.181
```

But that doesn't seem to be a problem.

Reply:
Thanks to you both, I learned something today :)